app/main.py

```diff
@@ -1,34 +1,90 @@
+        if start_row == end_row:
+            for col in range(start_col, end_col + 1):
+                decks.append(Deck(start_row, col))
+        elif start_col == end_col:
+            for row in range(start_row, end_row + 1):
+                decks.append(Deck(row, start_col))
+        return decks
+
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+
+    def fire(self, row: int, column: int) -> bool:
+        deck: Deck | None = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.is_alive = False
+            if all(not d.is_alive for d in self.decks):
+                self.is_drowned = True
+            return True
+        return False
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self,
+                 ships: list[tuple[tuple[int, int], tuple[int, int]]]
+                 ) -> None:
+        self.size = 10
+        self.field = [["~" for _ in range(self.size)]
+                      for _ in range(self.size)]
+        self.ships = [Ship(*ship) for ship in ships]
+        self.ship_cells = {}
+        self.hit_cells = set()
+        self._place_ships()
+
+    def _get_ship_cells(self, start: tuple[int, int],
+                        end: tuple[int, int]
+                        ) -> list[tuple[int, int]]:
+        cells = []
+        start_row, start_col = start
+        end_row, end_col = end
+        if start_row == end_row:
+            for col in range(start_col, end_col + 1):
+                cells.append((start_row, col))
+        elif start_col == end_col:
+            for row in range(start_row, end_row + 1):
+                cells.append((row, start_col))
+        return cells
+
+    def _place_ships(self) -> None:
+        for ship in self.ships:
+            for deck in ship.decks:
+                self.field[deck.row][deck.column] = u"\u25A1"
+
+    def fire(self, location: tuple[int, int]) -> str:
+        row, col = location
+        if self.field[row][col] == "~":
+            return "Miss!"
+        for ship in self.ships:
+            if ship.fire(row, col):
+                self.field[row][col] = "*"
+                if ship.is_drowned:
+                    for deck in ship.decks:
+                        self.field[deck.row][deck.column] = "x"
+                    return "Sunk!"
+                return "Hit!"
+        return "Miss!"
```

Comment:
The `fire` method in the `Battleship` class has a redundant `return` statement. According to the checklist, you should avoid using `return` statements that don't do anything. In this case, the `return` statement on line 86 is not needed because the function will return `None` by default if it doesn't hit any of the `return` statements in the `if` conditions. Here's how you can rewrite this method:

```python
    def fire(self, location: tuple[int, int]) -> str:
        row, col = location
        if self.field[row][col] == "~":
            return "Miss!"
        for ship in self.ships:
            if ship.fire(row, col):
                self.field[row][col] = "*"
                if ship.is_drowned:
                    for deck in ship.decks:
                        self.field[deck.row][deck.column] = "x"
                    return "Sunk!"
                return "Hit!"
```
This version of the method is cleaner and more Pythonic.